graphql.go

```diff
@@ -0,0 +1,148 @@
+package goshopify
+
+import (
+	"time"
+)
+
+// GraphQLService is an interface to interact with the graphql endpoint
+// of the Shopify API
+// See https://shopify.dev/docs/admin-api/graphql/reference
+type GraphQLService interface {
+	Query(string, interface{}, interface{}) error
+}
+
+// GraphQLServiceOp handles communication with the graphql endpoint of
+// the Shopify API.
+type GraphQLServiceOp struct {
+	client *Client
+}
+
+type graphQLResponse struct {
+	Data       interface{}       `json:"data"`
+	Errors     []graphQLError    `json:"errors"`
+	Extensions *graphQLExtension `json:"extensions"`
+}
+
+type graphQLExtension struct {
+	Cost GraphQLCost `json:"cost"`
+}
+
+// GraphQLCost represents the cost of the graphql query
+type GraphQLCost struct {
+	RequestedQueryCost int                   `json:"requestedQueryCost"`
+	ActualQueryCost    *int                  `json:"actualQueryCost"`
+	ThrottleStatus     GraphQLThrottleStatus `json:"throttleStatus"`
+}
+
+// GraphQLThrottleStatus represents the status of the shop's rate limit points
+type GraphQLThrottleStatus struct {
+	MaximumAvailable   float64 `json:"maximumAvailable"`
+	CurrentlyAvailable float64 `json:"currentlyAvailable"`
+	RestoreRate        float64 `json:"restoreRate"`
+}
+
+type graphQLError struct {
+	Message    string                 `json:"message"`
+	Extensions *graphQLErrorExtension `json:"extensions"`
+	Locations  []graphQLErrorLocation `json:"locations"`
+}
+
+type graphQLErrorExtension struct {
+	Code          string
+	Documentation string
+}
+
+const (
+	graphQLErrorCodeThrottled = "THROTTLED"
+)
+
+type graphQLErrorLocation struct {
+	Line   int `json:"line"`
+	Column int `json:"column"`
+}
+
+// Query creates a graphql query against the Shopify API
+// the "data" portion of the response is unmarshalled into resp
+func (s *GraphQLServiceOp) Query(q string, vars, resp interface{}) error {
+	data := struct {
+		Query     string      `json:"query"`
+		Variables interface{} `json:"variables"`
+	}{
+		Query:     q,
+		Variables: vars,
+	}
+
+	attempts := 0
+
+	for {
+		gr := graphQLResponse{
+			Data: resp,
+		}
+
+		err := s.client.Post("graphql.json", data, &gr)
+		// internal attempts count towards outer total
+		attempts += s.client.attempts
+		s.client.attempts = attempts
```

Comment:
I get why we want line 84 but why do you need copy the total attempts back to the client? 

`attempts` isn't even used for anything, it's just incremented and the tests use it to see if it matches retries. I have no context but the original PR and that already had it like this which is odd (unless it was only ever meant to be a helper for the tests)

Reply:
From what I can tell, it was only for the tests.

So it is less confusing, I moved attempts to a return value of the method. I'm not sure anyone will have a use for the value, but it allows the tests to run without some hijacking of other variables

Reply:
`I'm not sure anyone will have a use for the value` - I agree so let's not inconvenience everyone just for the sake of the tests. Can you remove attempts as the return value? 
If we really care about testing the retries then we could pass out own httpClient and count the number of times a UR is requested. Or even simpler, use httpmock.GetCallCountInfo() to get the number of calls for each URL.

Reply:
Good call, `GetCallCountInfo` worked a treat